pkg/driver/executor/executor.go

```diff
@@ -376,72 +355,109 @@ func (ex *Executor) resolveNetworkIDsForPodNetwork() (sets.String, error) {
 	return podNetworkIDs, nil
 }
 
-// DeleteMachine deletes a server based on the supplied ID or name. The machine must have the cluster/role tags for any operation to take place.
-// If providerID is specified it takes priority over the machineName. If no providerID is specified, DeleteMachine will
-// try to resolve the machineName to an appropriate server ID.
+// DeleteMachine deletes a server based on the supplied machineName. If a providerID is supplied it is used instead of the
+// machineName to locate the server.
 func (ex *Executor) DeleteMachine(ctx context.Context, machineName, providerID string) error {
 	var (
-		err    error
 		server *servers.Server
+		err    error
 	)
 
-	if isEmptyString(pointer.StringPtr(providerID)) {
-		server, err = ex.getMachineByName(ctx, machineName)
+	if !isEmptyString(pointer.StringPtr(providerID)) {
+		serverID := decodeProviderID(providerID)
+		server, err = ex.getMachineByID(ctx, serverID)
 	} else {
-		server, err = ex.getMachineByProviderID(ctx, providerID)
+		server, err = ex.getMachineByName(ctx, machineName)
 	}
-	if err != nil {
-		if errors.Is(err, ErrNotFound) {
-			return nil
-		}
+	if err != nil && !errors.Is(err, ErrNotFound) {
 		return err
 	}
 
-	klog.V(1).Infof("deleting server [ID=%s]", server.ID)
-	if err := ex.Compute.DeleteServer(server.ID); err != nil {
-		return err
-	}
+	if err == nil {
+		klog.V(1).Infof("deleting server [Name=%s, ID=%s]", server.Name, server.ID)
+		if err := ex.Compute.DeleteServer(server.ID); err != nil {
+			return err
+		}
+
+		if err = ex.waitForStatus(server.ID, nil, []string{client.ServerStatusDeleted}, 300); err != nil {
+			return fmt.Errorf("error while waiting for server [ID=%q] to be deleted: %v", server.ID, err)
+		}
 
-	if err = ex.waitForStatus(server.ID, nil, []string{client.ServerStatusDeleted}, 300); err != nil {
-		return fmt.Errorf("error while waiting for server [ID=%q] to be deleted: %v", server.ID, err)
 	}
 
-	if !isEmptyString(ex.Config.Spec.SubnetID) {
+	if ex.isUserManagedNetwork() {
 		return ex.deletePort(ctx, machineName)
 	}
 
 	return nil
 }
 
+func (ex *Executor) getOrCreatePort(_ context.Context, machineName string) (string, error) {
+	var (
+		err              error
+		securityGroupIDs []string
+	)
+
+	portID, err := ex.Network.PortIDFromName(machineName)
+	if err == nil {
+		klog.V(2).Infof("found port [Name=%q, ID=%q]... skipping creation", machineName, portID)
+		return portID, nil
+	}
+	if err != nil && !client.IsNotFoundError(err) {
```

Comment:
Same here. The `err` will always be not nil after checking above `err == nil` and returning in this case.

Reply:
Here we can just remove the err != nil since the function returns inside the `err==nil` block.
Unless you also prefer to restructure with elses:

```
if err == nil {
} else if !client.IsNotFoundError(err) {
}